testcontainers-dapr/src/main/java/io/dapr/testcontainers/DaprContainer.java

```diff
@@ -0,0 +1,447 @@
+    }
+
+    public String getVersion() {
+      return version;
+    }
+  }
+
+  private static final int DAPRD_HTTP_PORT = 3500;
+  private static final int DAPRD_GRPC_PORT = 50001;
+  private final Set<Component> components = new HashSet<>();
+  private final Set<Subscription> subscriptions = new HashSet<>();
+  private String appName;
+  private Integer appPort = 8080;
+  private DaprLogLevel daprLogLevel = DaprLogLevel.INFO;
+  private String appChannelAddress = "localhost";
+  private String placementService = "placement";
+  private static final DockerImageName DEFAULT_IMAGE_NAME = DockerImageName.parse("daprio/daprd");
+  private Yaml yaml;
+  private DaprPlacementContainer placementContainer;
+  private String placementDockerImageName = "daprio/placement";
+  private boolean shouldReusePlacement;
+
+  /**
+   * Creates a new Dapr container.
+   * @param dockerImageName Docker image name.
+   */
+  public DaprContainer(DockerImageName dockerImageName) {
+    super(dockerImageName);
+    dockerImageName.assertCompatibleWith(DEFAULT_IMAGE_NAME);
+    // For susbcriptions the container needs to access the app channel
+    withAccessToHost(true);
+    // Here we don't want to wait for the Dapr sidecar to be ready, as the sidecar
+    // needs to
+    // connect with the application for susbcriptions
+
+    withExposedPorts(DAPRD_HTTP_PORT, DAPRD_GRPC_PORT);
+
+    DumperOptions options = new DumperOptions();
+    options.setDefaultFlowStyle(DumperOptions.FlowStyle.BLOCK);
+    options.setPrettyFlow(true);
+
+    Representer representer = new YamlRepresenter(options);
+    representer.addClassTag(MetadataEntry.class, Tag.MAP);
+    this.yaml = new Yaml(representer);
+  }
+
+  /**
+   * Creates a new Dapr container.
+   * @param image Docker image name.
+   */
+  public DaprContainer(String image) {
+    this(DockerImageName.parse(image));
+  }
+
+  public Set<Component> getComponents() {
+    return components;
+  }
+
+  public Set<Subscription> getSubscriptions() {
+    return subscriptions;
+  }
+
+  public DaprContainer withAppPort(Integer port) {
+    this.appPort = port;
+    return this;
+  }
+
+  public DaprContainer withPlacementService(String placementService) {
+    this.placementService = placementService;
+    return this;
+  }
+
+  public DaprContainer withAppName(String appName) {
+    this.appName = appName;
+    return this;
+  }
+
+  public DaprContainer withDaprLogLevel(DaprLogLevel daprLogLevel) {
+    this.daprLogLevel = daprLogLevel;
+    return this;
+  }
+
+  public DaprContainer withSubscription(String name, String pubSubName, String pubSubTopic, String route) {
+    subscriptions.add(new Subscription(name, pubSubName, pubSubTopic, route));
+    return this;
+  }
+
+  public DaprContainer withComponent(Component component) {
+    components.add(component);
+    return this;
+  }
+
+  public DaprContainer withComponent(String name, String type, String version, List<MetadataEntry> metadataEntries) {
```

Comment:
Delete this method. It adds too little value compared to instantiating a component first.